File: backend/app/services/azure_search.py
import requests
import json
import time
import pandas as pd
from typing import Dict

class AzureAIFoundryRAG:
    def __init__(self, search_service_name: str, search_admin_key: str, search_index_name: str):
        self.search_service_name = search_service_name
        self.search_admin_key = search_admin_key
        self.search_index_name = search_index_name
        self.search_endpoint = f"https://{search_service_name}.search.windows.net"

    def create_search_index(self):
        """Create Azure AI Search index for RAG"""
        index_definition = {
            "name": self.search_index_name,
            "fields": [
                {"name": "id", "type": "Edm.String", "key": True, "searchable": False},
                {"name": "content", "type": "Edm.String", "searchable": True, "analyzer": "en.microsoft"},
                {"name": "title", "type": "Edm.String", "searchable": True, "filterable": True},
                {"name": "metadata", "type": "Edm.String", "searchable": False},
                {"name": "vector", "type": "Collection(Edm.Single)", "dimensions": 1536, "vectorSearchProfile": "vector-profile"}
            ],
            "vectorSearch": {
                "profiles": [{"name": "vector-profile", "algorithm": "hnsw-algorithm"}],
                "algorithms": [{"name": "hnsw-algorithm", "kind": "hnsw"}]
            }
        }

        headers = {"Content-Type": "application/json", "api-key": self.search_admin_key}
        url = f"{self.search_endpoint}/indexes/{self.search_index_name}?api-version=2023-11-01"
        response = requests.put(url, headers=headers, json=index_definition)
        response.raise_for_status()
        return True

    def index_powerbi_data(self, dataframes_dict: Dict[str, pd.DataFrame]):
        """Index Power BI data for RAG retrieval"""
        documents = []

        for query_name, df in dataframes_dict.items():
            content = df.to_string(index=False)

            doc = {
                "@search.action": "upload",
                "id": f"powerbi_{query_name}_{int(time.time())}",
                "content": content,
                "title": f"Power BI Data - {query_name}",
                "metadata": json.dumps({
                    "source": "PowerBI",
                    "query": query_name,
                    "timestamp": time.time(),
                    "row_count": len(df),
                    "columns": list(df.columns)
                })
            }
            documents.append(doc)

        headers = {"Content-Type": "application/json", "api-key": self.search_admin_key}
        upload_payload = {"value": documents}

        url = f"{self.search_endpoint}/indexes/{self.search_index_name}/docs/index?api-version=2023-11-01"
        response = requests.post(url, headers=headers, json=upload_payload)
        response.raise_for_status()
        return True

File: backend/src/services/azureSearch.js
// Import the necessary clients and classes from the Azure Search Documents library
import { SearchClient, SearchIndexClient, AzureKeyCredential } from '@azure/search-documents';
// Import the application configuration
import config from '../config/config.js';

/**
 * @class AzureSearchService
 * @description Manages interactions with the Azure AI Search service.
 * This class provides methods to create a search index and to upload
 * documents to the index, making them searchable.
 */
class AzureSearchService {
  constructor() {
    // The endpoint for the Azure AI Search service
    const endpoint = `https://${config.azureSearch.serviceName}.search.windows.net`;
    // The credential for authenticating with the service
    const credential = new AzureKeyCredential(config.azureSearch.adminKey);

    // Client for managing search indexes
    this.indexClient = new SearchIndexClient(endpoint, credential);
    // Client for interacting with a specific index
    this.searchClient = new SearchClient(endpoint, config.azureSearch.indexName, credential);
  }

  /**
   * @method createSearchIndex
   * @description Creates a new search index in the Azure AI Search service.
   * This method defines the schema for the index, including fields for vector search,
   * and then creates the index if it doesn't already exist.
   * @returns {Promise<boolean>} A boolean indicating if the index was created successfully.
   */
  async createSearchIndex() {
    // The definition of the search index
    const index = {
      name: config.azureSearch.indexName,
      fields: [
        { name: 'id', type: 'Edm.String', key: true, searchable: false },
        { name: 'content', type: 'Edm.String', searchable: true, analyzer: 'en.microsoft' },
        { name: 'title', type: 'Edm.String', searchable: true, filterable: true },
        { name: 'metadata', type: 'Edm.String', searchable: false },
        { name: 'vector', type: 'Collection(Edm.Single)', dimensions: 1536, vectorSearchProfile: 'vector-profile' },
      ],
      vectorSearch: {
        profiles: [{ name: 'vector-profile', algorithm: 'hnsw-algorithm' }],
        algorithms: [{ name: 'hnsw-algorithm', kind: 'hnsw' }],
      },
    };

    // Create the index, ignoring errors if it already exists
    await this.indexClient.createOrUpdateIndex(index);
    return true;
  }

  /**
   * @method indexPowerBIData
   * @description Indexes data from Power BI into the Azure AI Search index.
   * This method takes a dictionary of dataframes, converts them into documents
   * that match the index schema, and then uploads them to the search index.
   * @param {Object.<string, Array<object>>} dataframes - A dictionary where keys are query names and values are the data.
   * @returns {Promise<boolean>} A boolean indicating if the data was indexed successfully.
   */
  async indexPowerBIData(dataframes) {
    const documents = [];

    // Iterate over each dataframe and create a document for it
    for (const [queryName, df] of Object.entries(dataframes)) {
      const content = JSON.stringify(df); // Convert the dataframe to a string

      const doc = {
        '@search.action': 'upload',
        id: `powerbi_${queryName}_${Date.now()}`,
        content: content,
        title: `Power BI Data - ${queryName}`,
        metadata: JSON.stringify({
          source: 'PowerBI',
          query: queryName,
          timestamp: new Date().toISOString(),
          rowCount: df.length,
          columns: Object.keys(df[0] || {}), // Get column names from the first row
        }),
      };
      documents.push(doc);
    }

    // Upload the documents to the search index
    await this.searchClient.uploadDocuments(documents);
    return true;
  }
}

// Export the AzureSearchService class
export default AzureSearchService;
